boards: arm: Add support for SiLabs EFM32 Giant Gecko GG11 STK

Oanerer opened the pull request:
This PR adds initial support for the Silicon Labs EFM32 Giant Gecko GG11 Starter Kit (EFM32GG-STK3701A).

Peripherals supported for this board are currently NVIC, SysTick, GPIO, all UARTs/USARTs/LEUARTs, I2C, Flash, Counter and Ethernet. ADC and Watchdog support will be added as soon as their respective PRs are merged. 

The following samples were tested on this board:
* hello_world
* nvs
* counter
* big_http_download
... and a variety of other self written applications.

Note: this PR depends on a PR [#2](https://github.com/zephyrproject-rtos/hal_silabs/pull/2) for hal_silabs. A west.yml entry will be added to the PR when it has been merged.

Commit:
soc/arm/silabs_exx32: Add support for EFM32GG11B SoC

This commit adds support for the Silicon Labs EFM32GG11B
series of SoCs.

Signed-off-by: Oane Kingma <o.kingma@interay.com>

Commit:
boards/arm: Add initial support for SiLabs Giant Gecko GG11 STK

This commit adds initial support for the Silicon Labs EFM32
Giant Gecko GG11 StarterKit.

Features supported for now are NVIC, SysTick, GPIO, Flash,
Counter, I2C, UART and Ethernet. Support for Watchdog and
ADC will follow as soon as their respective PRs are merged.

Signed-off-by: Oane Kingma <o.kingma@interay.com>

Commit:
drivers/ethernet: Add support for SiLabs Giant Gecko GG11

Ethernet MAC present in Silicon Labs EFM32GG11B4xx and
EFM32GG11B8xx SoCs.

DMA based driver with support for link up/down detection.

Signed-off-by: Oane Kingma <o.kingma@interay.com>

Commit:
drivers/serial: Add more USART ports for Silabs EFM32GG11B

Silicon Labs EFM32GG11B SoCs have more USART ports than
currently supported by the driver. Driver expanded with
support for USART4 and USART5.

Signed-off-by: Oane Kingma <o.kingma@interay.com>

dbkinder commented on boards/arm/efm32gg_stk3701a/doc/index.rst:
```diff
@@ -0,0 +1,203 @@
+The EFM32GG11 SoC has nine GPIO controllers (PORTA to PORTI), all of which are
+currently enabled for the EFM32GG-STK3701A board.
+
+In the following table, the column **Name** contains pin names. For example, PE2
```
Interesting that you picked an example "PE2" that's not an entry in the table.

Oanerer replied:
True, I'm afraid this is a case of 'copy/paste, no time to waste'. Modified to PE1, which *is* actually used on the board.

dbkinder commented on boards/arm/efm32gg_stk3701a/doc/index.rst:
```diff
@@ -0,0 +1,203 @@
+Serial Port
+===========
+
+The EFM32GG11 SoC has six USARTs, two UARTs and two Low Energy UARTs (LEUART).
```
Do we need to know which UART is assigned to the console?

Oanerer replied:
Added (USART4 in this case)

dbkinder commented on boards/arm/efm32gg_stk3701a/doc/index.rst:
```diff
@@ -0,0 +1,203 @@
+should see a USB connection which exposes a mass storage device(STK3701A).
+Copy the generated zephyr.bin to the STK3701A drive.
+
+Use a USB-to-UART converter such as an FT232/CP2102 to connect to the UART on the
```
Do you need to clarify which pins on the header to connect to?

Oanerer replied:
This is definately a case of 'copy/paste'. The console is assigned to USART4, which is exposed through the USB connector along with the mass storage device. So corrected that and removed any mention of the USB-to-UART converter.

Oanerer reviewed (commented):
David, thanks for your time reviewing. If you feel anything is lacking in general with regards to the documentation please let me know!

dbkinder reviewed (approved):
docs LGTM, thanks.

jukkar commented on drivers/ethernet/eth_gecko.c:
```diff
@@ -0,0 +1,680 @@
+#define LOG_LEVEL CONFIG_ETHERNET_LOG_LEVEL
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
```
This is old way of setting up logging. Instead of this, you could just write
```
#include <logging/log.h>
LOG_MODULE_REGISTER(eth_gecko, CONFIG_ETHERNET_LOG_LEVEL);
```

Oanerer replied:
Changed

jukkar commented on drivers/ethernet/eth_gecko.c:
```diff
@@ -0,0 +1,680 @@
+		res = net_recv_data(dev_data->iface, rx_frame);
+		if (res < 0) {
+			LOG_ERR("Failed to enqueue frame into RX queue: %d",
+			res);
```
Indentation wrong here

Oanerer replied:
Corrected

jukkar commented on drivers/ethernet/eth_gecko.c:
```diff
@@ -0,0 +1,680 @@
+		if (res < 0) {
+			LOG_ERR("Failed to enqueue frame into RX queue: %d",
+			res);
+			/*eth_stats_update_errors_rx(dev_data->iface);*/
```
You mentioned in the beginning of the file that statistics are not supported. To me it looks quite trivial task to support that as you only need the interface pointer for it.

Oanerer replied:
I actually hadn't taken a good look yet at what supporting statistics would require. I uncommented line 232 in the section above, is that enough for this driver or are there other things I should support? If so, could you point me to a driver that I can use as a starting point?

jukkar replied:
It seems that currently all the existing drivers just have `eth_stats_update_errors_rx(context->iface);` to update the RX error count. So here you could just uncomment this line and it should work just fine. Just add `#include <ethernet/eth_stats.h>` at the beginning of the file. Currently all the other Ethernet stats are updated by the generic Ethernet L2 driver.

Oanerer replied:
I see, thank you! I've uncommented `eth_stats_update_errors_rx(context->iface);` and included `<ethernet/eth_stats.h>`.

jukkar commented on drivers/ethernet/phy_gecko.c:
```diff
@@ -0,0 +1,287 @@
+#define LOG_LEVEL CONFIG_ETHERNET_LOG_LEVEL
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
```
This could also shortened to
```
#include <logging/log.h>
LOG_MODULE_REGISTER(eth_gecko_phy, CONFIG_ETHERNET_LOG_LEVEL);
```
or you could just use the eth_gecko module too, in which case this could be written as
```
#include <logging/log.h>
LOG_MODULE_DECLARE(eth_gecko, CONFIG_ETHERNET_LOG_LEVEL);
```

Oanerer replied:
Changed to the first suggestion

jukkar commented on drivers/ethernet/phy_gecko.c:
```diff
@@ -0,0 +1,287 @@
+	}
+
+	LOG_INF("common abilities: speed %s Mb, %s duplex",
+		*status & ETH_NETWORKCFG_FULLDUPLEX ? "full" : "half",
```
The prints seem to be wrong order, speed should be first followed by duplex setting.

Oanerer replied:
Oops, corrected

jukkar reviewed (changes requested):
Minor nit picks for the ethernet driver

Oanerer reviewed (commented):
Thanks Jukka for your time reviewing, much appreciated!

jukkar reviewed (approved):
+1 to Ethernet driver

jhedberg commented:
@Oanerer you'll need to address the failed Codeowners and Shippable checks before this can be merged. Any update on this?

Oanerer commented:
> 
> 
> @Oanerer you'll need to address the failed Codeowners and Shippable checks before this can be merged. Any update on this?
Thanks for asking!

1. The checkpatch warnings are mostly lines over 80 characters, but I don't know if that is an issue?
2. Same question for the `WARNING:UNDOCUMENTED_DT_STRING` warning; is that an issue? If so, how should I proceed?
3. Shippable fails probably because this PR depends on PR [#2](https://github.com/zephyrproject-rtos/hal_silabs/pull/2) in `hal_silabs`. If someone can merge that, I will also add a `west.yml` entry to that to this PR.
4. I will add a `CODEOWNERS` entry for the files along with the `west.yml` entry.

jhedberg commented:
@Oanerer 1 and 2 are probably not blockers. Thanks for letting me know about the dependent hal_silabs PR (I should have noticed the reference to it in the description).

jhedberg commented:
@Oanerer I've gone and merged the HAL PR now

Oanerer commented:
> 
> 
> @Oanerer I've gone and merged the HAL PR now

Much appreciated! I updated `west.yml` accordingly and added an entry to `CODEOWNERS`. Hopefully shippable passes now also.

jhedberg reviewed (changes requested):
Some minor cosmetic issues I noticed, which you may want to fix

Oanerer commented on drivers/ethernet/eth_gecko_priv.h:
```diff
@@ -0,0 +1,143 @@
+
+#define ETH_BUF_ALIGNMENT     (16)
+
+#define ETH_DESC_ALIGNMENT    (4)
```
Sure, removed them. 

As a force of habit I tend to add them as I personally rather have some redundant ones than one too little that causes a weird bug.

Oanerer commented on drivers/ethernet/eth_gecko_priv.h:
```diff
@@ -0,0 +1,143 @@
+#define ETH_TX_UNDERRUN       (0x1UL<<28)
+#define ETH_TX_EXHAUSTED      (0x1UL<<27)
+#define ETH_TX_NO_CRC         (0x1UL<<16)
+#define ETH_TX_LAST           (0x1UL<<15)
```
Modified these to the BIT() version and thanks for pointing it it's existence!

Oanerer commented on drivers/ethernet/eth_gecko_priv.h:
```diff
@@ -0,0 +1,143 @@
+#define ETH_RX_CFI            (0x1UL<<16)
+#define ETH_RX_EOF            (0x1UL<<15)
+#define ETH_RX_SOF            (0x1UL<<14)
+#define ETH_RX_OFFSET         (0x3UL<<12)
```
Done for 1 bit only values, left `ETH_RX_VLAN_PRIORITY` and `ETH_RX_OFFSET` intact.

Oanerer commented on drivers/ethernet/eth_gecko_priv.h:
```diff
@@ -0,0 +1,143 @@
+#define ETH_RX_DISABLE()      (eth->NETWORKCTRL &= ~ETH_NETWORKCTRL_ENBRX)
+
+#define ETH_TX_ENABLE()       (eth->NETWORKCTRL |= ETH_NETWORKCTRL_ENBTX)
+#define ETH_TX_DISABLE()      (eth->NETWORKCTRL &= ~ETH_NETWORKCTRL_ENBTX)
```
I went for the latter approach, using the `eth` base address as the parameter for the macro.

Oanerer commented on drivers/ethernet/phy_gecko.c:
```diff
@@ -0,0 +1,284 @@
+/* Enable MDIO serial bus between MAC and PHY. */
+static void mdio_bus_disable(ETH_TypeDef *eth)
+{
+	eth->NETWORKCTRL &= (~ETH_NETWORKCTRL_MANPORTEN);
```
Removed

Oanerer commented:
> 
> 
> Some minor cosmetic issues I noticed, which you may want to fix

Thank you, the suggested changes have been implemented.

jhedberg reviewed (approved):
Looks ok to me, but it'd be good to get some more reviews from the assigned reviewers.

ioannisg commented on boards/arm/efm32gg_stk3701a/Kconfig:
```diff
@@ -0,0 +1,18 @@
+config BOARD_INIT_PRIORITY
+	int
+
+	# omit prompt to signify a "hidden" option
```
nit; can you remove this comment line?

Oanerer replied:
Minor nit fixed, thanks

Oanerer commented:
So... out of the blue Shippable fails on `SOC_FAMILY_EXX32` which still seems present in master and used by all other Gecko boards! How to proceed??

`error: SOC_FAMILY_EXX32 (defined at soc/arm/silabs_exx32/Kconfig:5) is assigned in a configuration file,
but is not directly user-configurable (has no prompt). It gets its value indirectly from other
symbols. See http://docs.zephyrproject.org/latest/reference/kconfig/CONFIG_SOC_FAMILY_EXX32.html
and/or look up SOC_FAMILY_EXX32 in the menuconfig/guiconfig interface. The Application Development
Primer, Setting Configuration Values, and Kconfig - Tips and Best Practices sections of the manual
might be helpful too.
`
